src/server/pfs/server/commit_set.go

```diff
@@ -0,0 +1,492 @@
+package server
+
+import (
+	"context"
+	"sort"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pachyderm/pachyderm/v2/src/client"
+	col "github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pfsdb"
+	"github.com/pachyderm/pachyderm/v2/src/internal/transactionenv/txncontext"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+	pfsserver "github.com/pachyderm/pachyderm/v2/src/server/pfs"
+)
+
+// returns CommitInfos in a commit set, topologically sorted.
+// A commit set will include all the commits that were created across repos for a run, along
+// with all of the commits that the run's commit's rely on (present in previous commit sets).
+func (d *driver) inspectCommitSetImmediateTx(txnCtx *txncontext.TransactionContext, commitSet *pfs.CommitSet, filterAliases bool) ([]*pfs.CommitInfo, error) {
+	var cis []*pfs.CommitInfo
+	if !filterAliases {
+		cs, err := pfsdb.CommitSetProvenance(txnCtx.SqlTx, commitSet.ID)
+		if err != nil {
+			return nil, err
+		}
+		for _, c := range cs {
+			ci := &pfs.CommitInfo{}
+			if err := d.commits.ReadWrite(txnCtx.SqlTx).Get(c, ci); err != nil {
+				return nil, err
+			}
+			cis = append(cis, ci)
+		}
+	}
+	ci := &pfs.CommitInfo{}
+	if err := d.commits.ReadWrite(txnCtx.SqlTx).GetByIndex(pfsdb.CommitsCommitSetIndex, commitSet.ID, ci, col.DefaultOptions(), func(string) error {
+		cis = append(cis, proto.Clone(ci).(*pfs.CommitInfo))
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return TopologicalSort(cis), nil
+}
+
+// TopologicalSort sorts a slice of commit infos topologically based on their provenance
+func TopologicalSort(cis []*pfs.CommitInfo) []*pfs.CommitInfo {
+	commits := make(map[string]*pfs.CommitInfo)
+	for _, ci := range cis {
+		commits[pfsdb.CommitKey(ci.Commit)] = ci
+	}
+	commitSubv := make(map[string][]string) // maps commit key -> []commit keys
+	sorted := make(map[string]struct{})
+	res := make([]*pfs.CommitInfo, 0)
+	// set up commitSubv
+	for _, ci := range commits {
+		provCount := 0
+		for _, p := range ci.DirectProvenance {
+			pKey := pfsdb.CommitKey(p)
+			if _, ok := commits[pKey]; ok {
+				commitSubv[pKey] = append(commitSubv[pKey], pfsdb.CommitKey(ci.Commit))
+				provCount++
+			}
+		}
+		if provCount == 0 {
+			res = append(res, ci)
+			sorted[pfsdb.CommitKey(ci.Commit)] = struct{}{}
+		}
+	}
+	for i := 0; i < len(cis); i++ {
+		k := pfsdb.CommitKey(res[i].Commit)
+		for _, c := range commitSubv[k] {
+			if _, ok := sorted[c]; !ok {
+				satisfied := true
+				ci := commits[c]
+				for _, p := range ci.DirectProvenance {
+					pk := pfsdb.CommitKey(p)
+					_, commitExists := commits[pk]
```

Comment:
This seems like a validation check as well. I would prefer this logic to just be if it is not in sorted, then add it to result.